typing/typecore.ml

```diff
@@ -3593,13 +3626,22 @@ and type_expect_
       begin_def ();
       let arg = type_exp env arg_expected_mode sarg in
       end_def ();
+      let sort =
+        match type_sort env arg.exp_type with
```

Comment:
As I think we've discussed since, the typedtree -> lambda translation wants to know sorts for lots of things, to avoid expensive layout checks, and this is one such case.

The suggestion to use `type_expect` is a clear improvement - done.